File: identity/src/androidTest/java/com/android/identity/StaticAuthDataTest.java
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.identity;

import androidx.core.util.Pair;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.SmallTest;

import org.junit.Assert;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import co.nstant.in.cbor.CborBuilder;
import co.nstant.in.cbor.builder.MapBuilder;
import co.nstant.in.cbor.model.DataItem;
import co.nstant.in.cbor.model.SimpleValue;
import co.nstant.in.cbor.model.UnicodeString;

@SuppressWarnings("deprecation")
@RunWith(AndroidJUnit4.class)
public class StaticAuthDataTest {

    @Test
    @SmallTest
    public void testStaticAuthData() throws Exception {
        // This test checks that the order of the maps in IssuerSignedItem is preserved
        // when using Utility.encodeStaticAuthData() and that no canonicalization takes
        // place.

        DataItem issuerSignedItem = new CborBuilder()
                .addMap()
                .put("random", new byte[] {0x50, 0x51, 0x52})
                .put("digestID", 42)
                .put("elementIdentifier", "dataElementName")
                .put(new UnicodeString("elementValue"), SimpleValue.NULL)
                .end()
                .build().get(0);
        byte[] encodedIssuerSignedItem = Util.cborEncode(issuerSignedItem);

        DataItem issuerSignedItem2 = new CborBuilder()
                .addMap()
                .put("digestID", 43)
                .put("random", new byte[] {0x53, 0x54, 0x55})
                .put("elementIdentifier", "dataElementName2")
                .put(new UnicodeString("elementValue"), SimpleValue.NULL)
                .end()
                .build().get(0);
        byte[] encodedIssuerSignedItem2 = Util.cborEncode(issuerSignedItem2);

        Map<String, List<byte[]>> issuerSignedMapping = new HashMap<>();
        issuerSignedMapping.put("org.namespace",
                Arrays.asList(encodedIssuerSignedItem, encodedIssuerSignedItem2));

        byte[] encodedIssuerAuth = Util.cborEncode(new CborBuilder()
                .addArray()
                    .addArray()
                    .end()
                .add(SimpleValue.NULL)
                .add(new byte[] {0x01, 0x02})
                .end()
                .build().get(0));

        byte[] staticAuthData = Utility.encodeStaticAuthData(issuerSignedMapping, encodedIssuerAuth);

        Assert.assertEquals(
                "{\n" +
                        "  \"digestIdMapping\": {\n" +
                        "    \"org.namespace\": [\n" +
                        "      24(<< {\n" +
                        "        \"random\": h'505152',\n" +
                        "        \"digestID\": 42,\n" +
                        "        \"elementIdentifier\": \"dataElementName\",\n" +
                        "        \"elementValue\": null\n" +
                        "      } >>),\n" +
                        "      24(<< {\n" +
                        "        \"digestID\": 43,\n" +
                        "        \"random\": h'535455',\n" +
                        "        \"elementIdentifier\": \"dataElementName2\",\n" +
                        "        \"elementValue\": null\n" +
                        "      } >>)\n" +
                        "    ]\n" +
                        "  },\n" +
                        "  \"issuerAuth\": [\n" +
                        "    [],\n" +
                        "    null,\n" +
                        "    h'0102'\n" +
                        "  ]\n" +
                        "}",
                CborUtil.toDiagnostics(
                        staticAuthData,
                        CborUtil.DIAGNOSTICS_FLAG_EMBEDDED_CBOR + CborUtil.DIAGNOSTICS_FLAG_PRETTY_PRINT));

        // Now check we can decode it
        Pair<Map<String, List<byte[]>>, byte[]> val = Utility.decodeStaticAuthData(staticAuthData);
        Map<String, List<byte[]>> digestIdMapping = val.first;

        // Check that the IssuerSignedItem instances are correctly decoded and the order
        // of the map matches what was put in above
        Assert.assertEquals(1, digestIdMapping.size());
        List<byte[]> list = digestIdMapping.get("org.namespace");
        Assert.assertNotNull(list);
        Assert.assertEquals(2, list.size());
        Assert.assertEquals(
                "{\n" +
                        "  \"random\": h'505152',\n" +
                        "  \"digestID\": 42,\n" +
                        "  \"elementIdentifier\": \"dataElementName\",\n" +
                        "  \"elementValue\": null\n" +
                        "}",
                CborUtil.toDiagnostics(list.get(0), CborUtil.DIAGNOSTICS_FLAG_PRETTY_PRINT));
        Assert.assertEquals(
                "{\n" +
                        "  \"digestID\": 43,\n" +
                        "  \"random\": h'535455',\n" +
                        "  \"elementIdentifier\": \"dataElementName2\",\n" +
                        "  \"elementValue\": null\n" +
                        "}",
                CborUtil.toDiagnostics(list.get(1), CborUtil.DIAGNOSTICS_FLAG_PRETTY_PRINT));
    }
}

File: multipaz/src/commonTest/kotlin/org/multipaz/mdoc/mso/StaticAuthDataTest.kt
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.multipaz.mdoc.mso

import org.multipaz.cbor.Bstr
import org.multipaz.cbor.Cbor
import org.multipaz.cbor.CborArray
import org.multipaz.cbor.CborMap
import org.multipaz.cbor.DataItem
import org.multipaz.cbor.DiagnosticOption
import org.multipaz.cbor.Simple
import org.multipaz.cbor.Tagged
import org.multipaz.cbor.buildCborArray
import org.multipaz.cbor.buildCborMap
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class StaticAuthDataTest {
    private fun createValidDigestIdMapping(): Map<String, List<ByteArray>> {
        val issuerSignedItemMetadata = buildCborMap {
            put("random", byteArrayOf(0x50, 0x51, 0x52))
            put("digestID", 42)
            put("elementIdentifier", "dataElementName")
            put("elementValue", Simple.NULL)
        }
        val isiMetadataBytes: DataItem = Tagged(24, Bstr(Cbor.encode(issuerSignedItemMetadata)))
        val encodedIsiMetadataBytes = Cbor.encode(isiMetadataBytes)
        val issuerSignedItemMetadata2 = buildCborMap {
            put("digestID", 43)
            put("random", byteArrayOf(0x53, 0x54, 0x55))
            put("elementIdentifier", "dataElementName2")
            put("elementValue", Simple.NULL)
        }
        val isiMetadata2Bytes: DataItem = Tagged(24, Bstr(Cbor.encode(issuerSignedItemMetadata2)))
        val encodedIsiMetadata2Bytes = Cbor.encode(isiMetadata2Bytes)
        val issuerSignedItemMetadata3 = buildCborMap {
            put("digestID", 44)
            put("random", byteArrayOf(0x53, 0x54, 0x55))
            put("elementIdentifier", "portrait")
            put("elementValue", Cbor.encode(Bstr(byteArrayOf(0x20, 0x21, 0x22, 0x23))))
        }
        val isiMetadata3Bytes: DataItem = Tagged(24, Bstr(Cbor.encode(issuerSignedItemMetadata3)))
        val encodedIsiMetadata3Bytes = Cbor.encode(isiMetadata3Bytes)
        val issuerSignedMapping = mutableMapOf<String, List<ByteArray>>()
        issuerSignedMapping["org.namespace"] = listOf(
            encodedIsiMetadataBytes,
            encodedIsiMetadata2Bytes,
            encodedIsiMetadata3Bytes
        )
        return issuerSignedMapping
    }

    private fun createValidIssuerAuth(): ByteArray {
        return Cbor.encode(
            buildCborArray {
                addArray()
                end()
                add(Simple.NULL)
                add(byteArrayOf(0x01, 0x02))
            }
        )
    }

    @Test
    fun testStaticAuthData() {
        // This test checks that the order of the maps in IssuerSignedItem is preserved
        // when and that no canonicalization takes place.
        val issuerSignedMapping = createValidDigestIdMapping()
        val encodedIssuerAuth = createValidIssuerAuth()
        val staticAuthData = StaticAuthDataGenerator(issuerSignedMapping, encodedIssuerAuth).generate()
        assertEquals(
            """{
  "digestIdMapping": {
    "org.namespace": [
      24(<< {
        "random": h'505152',
        "digestID": 42,
        "elementIdentifier": "dataElementName",
        "elementValue": null
      } >>),
      24(<< {
        "digestID": 43,
        "random": h'535455',
        "elementIdentifier": "dataElementName2",
        "elementValue": null
      } >>),
      24(<< {
        "digestID": 44,
        "random": h'535455',
        "elementIdentifier": "portrait",
        "elementValue": h'4420212223'
      } >>)
    ]
  },
  "issuerAuth": [
    [],
    null,
    h'0102'
  ]
}""",
            Cbor.toDiagnostics(
                staticAuthData,
                setOf(DiagnosticOption.EMBEDDED_CBOR, DiagnosticOption.PRETTY_PRINT)
            )
        )

        // Now check we can decode it
        val decodedStaticAuthData = StaticAuthDataParser(staticAuthData).parse()
        val digestIdMapping = decodedStaticAuthData.digestIdMapping

        // Check that the IssuerSignedItem instances are correctly decoded and the order
        // of the map matches what was put in above
        assertEquals(1, digestIdMapping.size.toLong())
        val list = digestIdMapping["org.namespace"]!!
        assertEquals(3, list.size.toLong())
        assertEquals(
            """24(<< {
  "random": h'505152',
  "digestID": 42,
  "elementIdentifier": "dataElementName",
  "elementValue": null
} >>)""",
            Cbor.toDiagnostics(
                list[0],
                setOf(DiagnosticOption.EMBEDDED_CBOR, DiagnosticOption.PRETTY_PRINT)
            )
        )
        assertEquals(
            """24(<< {
  "digestID": 43,
  "random": h'535455',
  "elementIdentifier": "dataElementName2",
  "elementValue": null
} >>)""",
            Cbor.toDiagnostics(
                list[1],
                setOf(DiagnosticOption.EMBEDDED_CBOR, DiagnosticOption.PRETTY_PRINT)
            )
        )
        val issuerAuth = decodedStaticAuthData.issuerAuth
        assertContentEquals(encodedIssuerAuth, issuerAuth)
    }

    @Test
    fun testStaticAuthDataExceptions() {
        assertFailsWith<IllegalArgumentException>() {
            StaticAuthDataGenerator(mapOf(), createValidIssuerAuth()).generate()
        }
    }
}
